HPC-8569 flow list resolver

manelcecs opened the pull request:
This PR add a new resolver for Flow Search.
Alongside add service for flows and nested entities linked with flows.
Also added test for resolver and util pagination methods.

:wq!

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,127 @@
+  async getFlows(args: GetFlowsArgs) {
+    const { models, conditions, offset, orderBy, limit } = args;
+
+    return await models!.flow.find({
```
Don't use non-null assertion operator!

Pl217 commented on src/domain-services/flows/strategy/impl/utils.ts:
```diff
@@ -0,0 +1,337 @@
+    }
+
+    const refDirectionMap = flowObjectsConditions.get(objectType);
+    if (!refDirectionMap!.has(direction)) {
```
Don't use non-null assertion operator!

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,177 @@
+  }
+
+  async getFlowIDsFromEntity(
+    models: Database,
```
`models` is unused

Pl217 commented on tools/hid_api_mock/package.json:
```diff
@@ -0,0 +1 @@
+/srv/example/package.json
```
What is this supposed to be?

Pl217 commented on tools/hid_api_mock/server.js:
```diff
@@ -0,0 +1 @@
+/srv/example/server.js
```
What is this supposed to be?

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (changes requested):
I didn't review this PR, but caught some parts during rebase, which I wrote about inline. Also, there's too many commits and lots of them are marked as temporary, but still there in the commit chain. Please rebase this work and make it easier to review.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Basic types for GraphQL resolvers

Upgrade hpc-api-core with latest develop version

manelcecs committed:
Flow resolver and Args

Needed nested properties GQL types

merge 6c551774dedb9dceb0f3d29084ea8959bcfa6d6b

manelcecs committed:
Flow search services

Nested needed services to complete Flow response

manelcecs committed:
Add 'only-flow-conditions-strategy' implementation

Add also utils to operate over flows

manelcecs committed:
Add 'search-flow-by-filters-strategy' implementation

Nested strategies to find flowIDs from different filters

Merge 5de9ca312a5fe37aeea198cbc6ba84bac352d3ea

manelcecs committed:
Add unit tests to cover search flows

Merge 141f445d9671365e135350796e120e0bdbf55ffc

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Delgee commented on src/domain-services/base-types.ts:
If we changing `Date` fields to `string` `deletedAt` field needs to change too.

Delgee commented on src/domain-services/location/graphql/types.ts:
```diff
@@ -25,7 +25,7 @@ export default class Location extends BaseType {
   name?: string;
 
-  @Field(() => Int)
+  @Field(() => Number)
```
Why this changed to `Number`?

manelcecs replied:
This is in order to not  depend (too much) on the types provided from 'type-graphql' since we are not sure if we want to keep it in a long term.

Delgee commented on src/domain-services/location/graphql/types.ts:
```diff
@@ -47,9 +47,48 @@ export default class Location extends BaseType {
   status?: LocationStatus; // Accidentally optional
 
-  @Field(() => Int, { nullable: true })
+  @Field(() => Number, { nullable: true })
```
Also this one

manelcecs replied:
Same above.
> This is in order to not depend (too much) on the types provided from 'type-graphql' since we are not sure if we want to keep it in a long term.

Delgee commented on src/domain-services/location/graphql/types.ts:
```diff
@@ -47,9 +47,48 @@ export default class Location extends BaseType {
 }
+
+@ObjectType()
+export class BaseLocation extends BaseTypeWithDirection {
```
Why is it named `BaseLocation` if it is supposed to be location with direction.

manelcecs replied:
Very true.

Delgee commented on src/domain-services/report-details/graphql/types.ts:
```diff
@@ -0,0 +1,46 @@
+  @Field(() => String, { nullable: true })
+  channel: string | null;
+}
+export type SystemID =
```
We should use types from `hpc-api-core` here. Instead of defining them inline.

manelcecs replied:
Done!

Delgee commented on src/domain-services/categories/model.ts:
```diff
@@ -0,0 +1,35 @@
+import { type Op } from '@unocha/hpc-api-core/src/db/util/conditions';
+
+export type CategoryGroup =
```
We also should use types from `hpc-api-core` here.

manelcecs replied:
Since its not used anywhere in the code - I'll be removed
In the future if needed, as you suggest, we should import from `hpc-api-core`.

Delgee commented on src/domain-services/flow-object/model.ts:
```diff
@@ -0,0 +1,30 @@
+export type FlowObject = InstanceOfModel<Database['flowObject']>;
+
+// Define the FlowObjectType as a runtime type
+export const FlowObjectType = t.union([
```
Use type from `hpc-api-core`.

manelcecs replied:
Updated!

Delgee commented on src/domain-services/flows/flow-search-service.ts:
```diff
@@ -0,0 +1,652 @@
+    isStandardFlows: boolean
+  ): ShortcutCategoryFilter[] | null {
+    const filters = [
+      { flag: isPendingFlows, category: 'Pending', id: 45 },
```
I don't think it is good idea to hardcode `id`.

manelcecs replied:
It was a workaround in order to reduce the amount of queries to perform and reduce the execution time but now it makes no sense.
I've updated it!

Delgee commented on src/domain-services/flows/flow-search-service.ts:
In old flow search there was a logic excluding flows with `Pending review` inactiveStatus. Looks like this logic is missing. Can it be replaced with just using `activeStatus === true`?

Delgee commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,335 @@
+      .queryBuilder()
+      .select('fChild.id', 'fChild.versionID')
+      .from('categoryRef as cr')
+      .where('cr.categoryID', 1252)
```
Can we change hardcoding `id`?

manelcecs replied:
Sure thing!

Delgee reviewed (changes requested):
Great job. I have only a few nitpicks.

Pl217 commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,335 @@
+    // because it does have a direct relation with flow
+    // and no direction
+    if (entity === 'externalReference') {
+      const entityList = await dbConnection
+        .queryBuilder()
+        .distinct('flowID', 'versionID', mappedOrderBy.column)
+        .from('externalReference')
+        .orderByRaw(orderByRaw);
```
Across this entire PR, you are violating a lot of the practices we've established around how DB models are used and how we query data. First, you shouldn't have exposed query builder at all. You should remove it and rewrite all the code to use `models` instead. Models already expose query builder, although we are very reluctant to use it and should only use it very rarely, when our models library cannot support such query. This would be an example how to use it:
```ts
await models.planTag.find({
  where: {
    [models.Cond.BUILDER]: (builder) => {
      builder.distinct(...);
    },
  },
});
```
However, the biggest reason we have our own model library is to have strong typing of our models. With using query builder from Knex directly, you lose that. And we don't want to give up on strong and strict typing.

What you should do instead is extend the model library to do what you want, or find a way around, for example, you can implement your own `DISTINCT` functionality by processing the "regular" results.

Then, you violate established practices even more further down in this method by using `JOIN`s. We have moved away from using joins a long time ago for 2 reasons:
1. Some of the joins become so complex that they take long time to execute, clogging the database and causing systemic slowdowns or plain crashes
2. We cannot have strong typing in Typescript when using joins. This is the biggest problem with all ORMs that are available in Node.js ecosystem. This is also one of the reasons we stopped using Sequelize (ORM) and moved to Knex (query builder).

I am using this code block just as an example of one place where you've violated established practices, but you should fix the problem elsewhere too.

Pl217 commented on src/domain-services/flows/flow-search-service.ts:
```diff
@@ -0,0 +1,652 @@
+      if (!flowWithVersion.has(flow.id)) {
+        flowWithVersion.set(flow.id, []);
+      }
+      const flowVersionIDs = flowWithVersion.get(flow.id)!;
```
Don't use non-null assertion operator! It should only be used in very rare circumstance where Typescript cannot infer the existence of a variable/method result. In this scenario, you could argue that `flowVersionIDs` can never be `undefined`, but you should have used [`getOrCreate`](https://github.com/UN-OCHA/hpc-api-core/blob/8a2a05fda25ac14af79ba5744f2f4a0cf6327765/src/util/index.ts#L205), which is a standardized method to do what you are doing here.

Since this isn't the only place where you're using non-null assertion operator, I will enable [`no-non-null-assertion` linting rule](https://typescript-eslint.io/rules/no-non-null-assertion/), to prevent such code from reaching code review in the future.

manelcecs replied:
Sure thing! I'm updating them all

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Delgee commented on src/domain-services/flows/model.ts:
```diff
@@ -0,0 +1,33 @@
+export type FlowModel = Database['flow'];
+export type FlowInstance = InstanceOfModel<FlowModel>;
+export type FlowWhere = Condition<FlowInstance>;
+export type FlowFieldsDefinition = FlowModel['_internals']['fields'];
```
Getting `FieldDefinition` from `_internals.fields` is incorrect, resulting in `FlowOrderByCond` will have column as `never` type. Can we change this? If it is difficult to get `FieldDefinition` from the model, can we address this in `hpc-api-core`. Due to this, there are a few instances of using `as key of` keyword usage. I don't think this is correct approach.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Delgee commented on src/domain-services/flow-object/flow-object-service.ts:
```diff
@@ -0,0 +1,150 @@
+    whereClauses: FlowObjectWhere,
+    orderBy?: FlowObjectOrderByCond
+  ): Promise<FlowObjectInstance[]> {
+    const distinctColumns = [
```
Let's declare types instead of forcing them.
```suggestion
    const distinctColumns: Array<keyof FlowObjectInstance> = [
      'flowID',
      'versionID',
    ];
```

Delgee commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,544 @@
+  async getFlows(args: IGetFlowsArgs): Promise<FlowInstance[]> {
+    const { models, orderBy, conditions, limit, offset } = args;
+
+    const distinctColumns = [
```
Let's declare types here too.
```suggestion
    const distinctColumns: Array<keyof FlowInstance> = ['id', 'versionID'];
```

Delgee commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,544 @@
+    switch (entity) {
+      case 'emergency': {
+        // Get emergency entities sorted
+        const column = orderBy.column as keyof InstanceOfModel<
```
We can use intersection type to handle this kind of issues. For example.
```ts
type FlowOrderByCommon = {
  order: SortOrder;
  direction?: EntityDirection;
};

export type FlowOrderBy = FlowOrderByCommon &
  (
    | {
        column: keyof FlowInstance;
        entity: 'flow';
      }
    | {
        column: keyof InstanceOfModel<Database['externalReference']>;
        entity: 'externalReference';
      }
    | {
        column: keyof InstanceOfModel<Database['emergency']>;
        entity: 'emergency';
      }
    | {
        entity: 'globalCluster';
        column: keyof InstanceOfModel<Database['globalCluster']>;
      }
    | {
        entity: 'governingEntity';
        column: keyof InstanceOfModel<Database['governingEntity']>;
      }
    | {
        entity: 'location';
        column: keyof InstanceOfModel<Database['location']>;
      }
    | {
        entity: 'organization';
        column: keyof InstanceOfModel<Database['organization']>;
      }
    | {
        entity: 'plan';
        column: keyof InstanceOfModel<Database['plan']>;
      }
    | {
        entity: 'usageYear';
        column: keyof InstanceOfModel<Database['usageYear']>;
      }
    | {
        entity: 'planVersion';
        column: keyof InstanceOfModel<Database['planVersion']>;
      }
    | {
        entity: 'project';
        column: keyof InstanceOfModel<Database['project']>;
      }
  );
```
We can also make it `io-ts` codec so we will have type checking in runtime. We can check if it is correct type or not on function https://github.com/UN-OCHA/hpc-api/blob/c6f882510e0154e65367fbdd4d0a2796239f586a/src/domain-services/flows/strategy/impl/utils.ts#L311

manelcecs replied:
Did you miss the last part of the comment? It looks like 😅

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Add runtime checks for 'input' columns

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Delgee commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,6 @@
+export const isValidKey = <T extends object, Q extends keyof T>(
+  key: keyof T
+): key is Q => {
+  const runtimeKeys = Object.keys({} as T) as (keyof T)[];
```
This won't work. On runtime every type will be stripped and this variable will be always empty array.

Delgee commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,6 @@
+  key: keyof T
+): key is Q => {
+  const runtimeKeys = Object.keys({} as T) as (keyof T)[];
+  return runtimeKeys.includes(key as keyof T);
```
So this will always give false.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Delgee commented on src/domain-services/flows/flow-service.ts:
```diff
@@ -0,0 +1,608 @@
+    // because it does have a direct relation with flow
+    // and no direction
+    if (entity === 'externalReference') {
+      const column = orderBy.column as keyof InstanceOfModel<
```
Missing runtime validation here

Delgee reviewed (approved):
Missing runtime checking on `externalReference`. Other than that looks good to me.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
